src/tools/rbd_mirror/image_map/ImageAction.h

```diff
@@ -0,0 +1,122 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+
+#include <map>
+#include <list>
+#include <string>
+
+#include "Types.h"
+#include "common/Mutex.h"
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+class ImageAction {
+public:
+  ImageAction();
+
+  enum ActionType {
+    ACTION_TYPE_ADD = 0,
+    ACTION_TYPE_REMOVE,
+    ACTION_TYPE_SHUFFLE,
+  };
+
+  enum SubActionType {
+    SUB_ACTION_TYPE_INVALID = 0,
+    SUB_ACTION_TYPE_MAP,
+    SUB_ACTION_TYPE_UNMAP,
+    SUB_ACTION_TYPE_PURGE,
+    SUB_ACTION_TYPE_COMPLETE,
+  };
+
+  struct Action {
+    ActionType action_type;
+
+    Context *on_map;
+    Context *on_unmap;
+    Context *on_purge;
+    Context *on_finish;
+
+    SubActionType sub_action_type; // currently executing sub action
+
+    Action(ActionType action_type)
```

Comment:
Nit: recommend moving this to private and creating three static factory methods for each action (i.e. ```create_remove```). Each factory method would take the params for its specific callbacks to be more explicit about what is needed.

Reply:
Make sense. In that case, do you have a preference if the returned object needs to be dynamically allocated to avoid unnecessary copying (one during the return, which mostly should be optimized by the compiler and another when the action is queued)?

`struct Action` is not huge for the copy to be expensive, so I'm guessing its ok.

Reply:
No need for allocating on the heap -- this is a tiny object. You can always use move semantics to transfer ownership. Of course, if the callbacks are going to be consolidated down to a fixed number for all actions due to other changes, perhaps this change isn't necessary.